workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/selection/VaultWatcherService.kt

```diff
@@ -0,0 +1,148 @@
+package com.r3.corda.sdk.token.workflow.selection
+
+import com.r3.corda.sdk.token.contracts.states.FungibleToken
+import com.r3.corda.sdk.token.contracts.types.IssuedTokenType
+import com.r3.corda.sdk.token.contracts.types.TokenType
+import com.r3.corda.sdk.token.workflow.selection.VaultWatcherService.Companion.LOG
+import net.corda.core.contracts.Amount
+import net.corda.core.contracts.StateAndRef
+import net.corda.core.node.AppServiceHub
+import net.corda.core.node.services.CordaService
+import net.corda.core.node.services.Vault
+import net.corda.core.node.services.vault.DEFAULT_PAGE_NUM
+import net.corda.core.node.services.vault.PageSpecification
+import net.corda.core.utilities.contextLogger
+import rx.Observable
+import java.lang.RuntimeException
+import java.security.PublicKey
+import java.time.Duration
+import java.util.concurrent.*
+
+val unlocker: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
+
+@CordaService
+class VaultWatcherService(vaultObserver: VaultObserver? = null) {
+
+    val cache: ConcurrentMap<PublicKey, ConcurrentMap<Class<*>, ConcurrentMap<String, ConcurrentMap<StateAndRef<FungibleToken<TokenType>>, Boolean>>>> = ConcurrentHashMap()
```

Comment:
Try using some types/classes!

Why not flatten the key to be the one you build in `processToken` ?  All the nested maps will be inefficient from a memory perspective.  Don't you just want `ConcurrentMap<BucketIdentifier, Bucket>`.

`processToken` becomes `fun generateBucketId(token: FungibleToken<*>): BucketIdentifier`

And for now `Bucket` just encapsulates the `ConcurrentMap<StateAndRef<FungibleToken<TokenType>>, Boolean>` - which I think should be changed anyway.... see other comments.